new mor application for heat transfert in eye fixes #2068

prudhomm opened the pull request:
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes?
- [ ] Have you successfully run the Feel++ testsuite with your changes locally?
- [ ] Have you written Doxygen comments in your contribution ?

-----

- closes #2068
- start work on #2081
- closes #2086
- closes #2093

prudhomm committed:
new mor application for heat transfert in eye fixes #2068

working now @vincentchabannes @lberti
@thomas-saigre

skip feelpp skip tests skip toolboxes skip python

thomas-saigre committed:
up begin of code for runnong the sensitivity analysis
for eye heat transfer model #2068
[ci skip]

thomas-saigre committed:
fix CMakeLists
[ci skip]

thomas-saigre committed:
tiny fix
[ci skip]

thomas-saigre committed:
fix warning at compilation
[ci skip]

thomas-saigre committed:
up : compute sobol indices with polynomial chaos
#2068

thomas-saigre committed:
fix computation of parameter from distribution
[ci skip]

thomas-saigre committed:
attemp to fix execution of python onlien script
/cc @prudhomm
[ci skip]

thomas-saigre committed:
fix execution
/cc @prudhomm @vincentchabannes

thomas-saigre committed:
set parameters as column of the dataframe
[ci skip]

thomas-saigre committed:
rename argument in function sampling
type -> samplingMode to avoid python errors
[ci skip]

thomas-saigre committed:
up (begin of) code to compute FEM and RB output
[ci skip]

thomas-saigre committed:
fix script to compare FEM and RB of the application
[ci skip]

thomas-saigre committed:
up crb_param
[ci skip]

thomas-saigre committed:
use output on Point O
[ci skip]

Co-authored-by: Christophe Prud'homme <christophe.prudhomme@cemosis.fr>

thomas-saigre committed:
up config files
[ci skip]

thomas-saigre committed:
up code to run toolbox with parameters
[ci skip]

thomas-saigre committed:
rename h_tau -> h_r for model
[ci skip]

prudhomm committed:
rename README [ci skip]

thomas-saigre committed:
fix E in formulation, and add h_r as a parameter
[ci skip]

prudhomm committed:
improve cmake preset #2082

- add mor-dbg
- fix CMAKE_INSTALL_PREFIX: adapt path to preset

thomas-saigre committed:
fix
[ci skip]

thomas-saigre committed:
up warning message in console
/cc @prudhomm
[ci skip]

thomas-saigre committed:
set h_r as a parameter
[ci skip]

thomas-saigre committed:
fix python script :
use P1 discretization for the toolbox
[ci skip]

thomas-saigre committed:
"fix" execution of script in parallel
[ci skip]

thomas-saigre committed:
fix execution of sensitivity analysis
#2068
/cc @prudhomm
[ci skip]

thomas-saigre committed:
start to work on outputs
[ci skip]

thomas-saigre committed:
work on #2086 :
add pluginlibdir as argument to loadDBPlugin, a new function is created
/cc @prudhomm @vincentchabannes

thomas-saigre committed:
Merge branch '2068-new-mor-application-for-heat-transfert-in-eye' of github.com:feelpp/feelpp into 2068-new-mor-application-for-heat-transfert-in-eye

thomas-saigre committed:
#2086 load plugin for python
-> wrap the function added in bf456afc
/cc @prudhomm

thomas-saigre committed:
fix brackets
[ci skip]

thomas-saigre committed:
add new python test for sampling

thomas-saigre committed:
refactor python code to compare FE and RB results

now we can choose to run only the toolbox code,
and also give results and parameters already computed

[ci skip]

thomas-saigre committed:
add function to plot results
[ci skip]

thomas-saigre committed:
Merge branch '2068-new-mor-application-for-heat-transfert-in-eye' of github.com:feelpp/feelpp into 2068-new-mor-application-for-heat-transfert-in-eye
[ci skip]

thomas-saigre committed:
tiny fixes for execution #2086

thomas-saigre committed:
attempt to use P2 discretisation
[ci skip]

thomas-saigre committed:
add a condition to find openturns #2093

/cc @prudhomm

[skip feelpp, skip toolboxes, skip tests, skip python]

thomas-saigre committed:
remove FATAL_ERROR flag #2093

[skip feelpp, skip toolboxes, skip tests, skip python]

thomas-saigre committed:
#2093 remove FATAL_ERROR flag for real
sorry !

[skip feelpp, skip toolboxes, skip tests, skip python]

thomas-saigre committed:
fix symbol for online execution

#2068

[ci skip]

prudhomm committed:
Merge

Update code in mor/apps/crbonlinerun.cpp, mor/examples/eye2brain/CMakeLists.txt, mor/examples/eye2brain/FunctionalChaos.hpp, mor/examples/eye2brain/README.md, and mor/examples/eye2brain/compFE-RB.py files.

- Modify crbonlinerun.cpp to loadDBPlugin from the given meta and pluginlibdir.
- Modify CMakeLists.txt to add a check for OpenTURNS, and conditionally build sensitivity_analysis.
- Modify FunctionalChaos.hpp to comment out the line that saves the validation plot as PDF and uncomment the line that saves the plot as a LaTex file.
- Add a README.md file to the example and a sample python script to run the example.
- Modify compFE-RB.py to enable it to compare the results of FE and RBM. In addition, it now plots the result and saves it as tex for future uses.

prudhomm committed:
add support for OpenTURNS #2091

Add support for OpenTURNS submodule and CMakeLists #2091

skip feelpp skip tests skip toolboxes

prudhomm committed:
add support for OpenTURNS #2091

install target OT feelpp_openturns

skip feelpp skip tests skip toolboxes

prudhomm committed:
update Openturns usage, make it conditional add support for OpenTURNS #2091

for app new mor application for heat transfert in eye #2068

skip feelpp skip tests skip toolboxes

prudhomm committed:
fix compilation with new mor application for heat transfert in eye #2068

using make conditional openturns dependency #2093 and add support for OpenTURNS #2091

skip feelpp skip tests skip toolboxes

prudhomm committed:
add support for OpenTURNS #2091

move commit
skip feelpp skip toolboxes skip tests

thomas-saigre committed:
up 2d application code :
- template for the class Eye2brain using diemnsion as template argument
- set 2 applications : one for the 2D and one for the 3D
- add corresponding config file for both 2D and 3D

/cc @prudhomm

[skip feelpp, skip toolboxes, skip python, skip tests]

thomas-saigre committed:
Merge branch '2068-new-mor-application-for-heat-transfert-in-eye' of github.com:feelpp/feelpp into 2068-new-mor-application-for-heat-transfert-in-eye

thomas-saigre committed:
fix online execution #2068

[ci skip]

thomas-saigre committed:
refactor crbonlinerun (and clean code) :
- get both outputs and error bound from implemented methods in CrbResults
- add errorbound in output table
- set floating precision from an option output_results.precision (default is still 6)

/cc @prudhomm

[ci skip]

thomas-saigre committed:
tiny fixes in names

[ci skip]

thomas-saigre committed:
fix buildPreset

[ci skip]

thomas-saigre committed:
fix loadMesh for P2 execution #2068

[ci skip]

/cc @prudhomm

thomas-saigre committed:
up template class for order

[skip feelpp, skip toolboxes, skip tests, skip python]
/cc @prudhomm

thomas-saigre committed:
up slurm script to run offline model

[ci skip]

thomas-saigre committed:
set custom target for cmake

[ci skip]

thomas-saigre committed:
refactor : using instead of typename

[ci skip]

thomas-saigre committed:
set options names from CMakeLists macro :

add OPT as argument of the macro,
if such argument is not given, that has the same behavior as before : mor_MODEL_OPT is used

[ci skip]

thomas-saigre committed:
add choice of the output with an option `--measure-index`

/cc @prudhomm
[skip feelpp, skip tests, skip toolboxes]

thomas-saigre committed:
fix for compilation

[skip feelpp, skip toolboxes, skip tests]

thomas-saigre committed:
add test script for eye2brain offline and online #2068

[skip feelpp, skip toolboxes, skip python, skip tests]

thomas-saigre committed:
fix load of plugin for online #2068

[skip feelpp, skip mor, skip tests, skip python]

thomas-saigre committed:
up slurm script to generated basis #2068
[ci skip]

thomas-saigre committed:
fix computation of the output with openturns
[ci skip]

thomas-saigre committed:
fix tests : use parallel for 3d and keep sequantial for 2d

and various other tiny fixes

[ci skip]

thomas-saigre committed:
update slurm script to generated all basis
-> now display the current output generated
#2068

[ci skip]

thomas-saigre committed:
add code to export output sample
and add doc
#2068

[ci skip]

prudhomm committed:
Merge branch 'develop' into 2068-new-mor-application-for-heat-transfert-in-eye

prudhomm commented:
@vincentchabannes could you review this PR please

vincentchabannes commented on mor/mor/feel/feelmor/parameterspace.hpp:
```diff
@@ -219,7 +219,10 @@ class ParameterSpace: public CommObject, public std::enable_shared_from_this<Par
                     this->operator()( it - paramNames.begin() ) = value;
                 else
+                {
+                    Feel::cout << tc::red << "Warning: " << tc::reset << value << " not in range [" << min_val << ", " << max_val << "]" << std::endl;
```
Warning Feel::cout use the environment worldcomm, and can not show the message if a sub communicator is used. LOG( Warning ) will do the job. 
However, here we can stop the application? with an exception thrown

thomas-saigre replied:
I added the display with `Feel::cout` because sometimes I forgot to check the log and spent some time to find that the issue what that I was setting parameters outside the range !
But yes we may want to throw an exception. @prudhomm what do you think ?

vincentchabannes commented on mor/mor/feel/feelmor/parameterspace.hpp:
```diff
@@ -231,7 +234,10 @@ class ParameterSpace: public CommObject, public std::enable_shared_from_this<Par
                 this->operator()(i) = value;
             else
+            {
+                Feel::cout << tc::red << "Warning: " << tc::reset << value << " not in range [" << min(i) << ", " << max(i) << "]" << std::endl;
```
idem

vincentchabannes commented on python/pyfeelpp-mor/feelpp/mor/crb.cpp:
```diff
@@ -240,7 +240,8 @@ PYBIND11_MODULE( _mor, m )
             { return db.loadDBMetaData( from, name ); },
             py::arg( "attribute" ) = crb::attributeToString( crb::attribute::last_modified ), py::arg( "data" ) = std::nullopt )
-        .def( "loadDBPlugin", static_cast<std::shared_ptr<CRBPluginAPI> ( CRBModelDB::* )( CRBModelDB::MetaData const&, std::string ) const>( &CRBModelDB::loadDBPlugin ), py::arg( "metadata" ), py::arg( "load" ) = "rb", "load DB plugin", pybind11::return_value_policy::reference );
+        .def( "loadDBPlugin", static_cast<std::shared_ptr<CRBPluginAPI> ( CRBModelDB::* )( CRBModelDB::MetaData const&, std::string ) const>( &CRBModelDB::loadDBPlugin ), py::arg( "metadata" ), py::arg( "load" ) = "rb", "load DB plugin", pybind11::return_value_policy::reference )
```
use std::string const&

vincentchabannes commented on python/pyfeelpp-mor/feelpp/mor/crb.cpp:
```diff
@@ -240,7 +240,8 @@ PYBIND11_MODULE( _mor, m )
             py::arg( "attribute" ) = crb::attributeToString( crb::attribute::last_modified ), py::arg( "data" ) = std::nullopt )
-        .def( "loadDBPlugin", static_cast<std::shared_ptr<CRBPluginAPI> ( CRBModelDB::* )( CRBModelDB::MetaData const&, std::string ) const>( &CRBModelDB::loadDBPlugin ), py::arg( "metadata" ), py::arg( "load" ) = "rb", "load DB plugin", pybind11::return_value_policy::reference );
+        .def( "loadDBPlugin", static_cast<std::shared_ptr<CRBPluginAPI> ( CRBModelDB::* )( CRBModelDB::MetaData const&, std::string ) const>( &CRBModelDB::loadDBPlugin ), py::arg( "metadata" ), py::arg( "load" ) = "rb", "load DB plugin", pybind11::return_value_policy::reference )
+        .def( "loadDBPlugin", static_cast<std::shared_ptr<CRBPluginAPI> ( CRBModelDB::* )( CRBModelDB::MetaData const&, std::string, std::string ) const>( &CRBModelDB::loadDBPlugin ), py::arg( "metadata" ), py::arg( "load" ) = "rb", py::arg( "pluginlibdir" ), "load DB plugin", pybind11::return_value_policy::reference );
```
use std::string const&

thomas-saigre committed:
use std::string const& in wrapper #2074

fix compilation due to wrong name of function in sensitivity_analysis

/cc @vincentchabannes
[skip feelpp, skip tests, skip toolboxes]

thomas-saigre committed:
up forgotten modifications in f39e31d696f58893cbf718a3d39be4a48fd880c9
about std::string const&
#2074

[skip feelpp, skip toolboxes, skip tests]

thomas-saigre commented:
@vincentchabannes I did insert the changes you requested, but it fails during the execution of the tests : https://buildkite.com/feelpp/feelpp/builds/4935#01887b08-2bcc-45e6-b101-a7c8e568064f
but I do not manage to find the cause of this failure : If I try locally, the tests pass well

thomas-saigre committed:
add compliant output
[skip feelpp, skip toolboxes, skip tests]

thomas-saigre committed:
crbonlinerun : add option to print results

add option output_results.print to select wheter or not the user want the results to be displayed in the shell
default to true

[skip feelpp, skip tests, skip toolboxes, skip python]

thomas-saigre committed:
up: throw exception when parameter is not in range

/cc @prudhomm @vincentchabannes

[skip feelpp, skip tests, skip toolboxes, skip python]

thomas-saigre committed:
Merge remote-tracking branch 'origin/develop' into 2068-new-mor-application-for-heat-transfert-in-eye

[skip feelpp, skip tests, skip toolboxes, skip python]

prudhomm commented on mor/mor/feel/feelmor/parameterspace.hpp:
```diff
@@ -220,8 +220,8 @@ class ParameterSpace: public CommObject, public std::enable_shared_from_this<Par
                 {
-                    Feel::cout << tc::red << "Warning: " << tc::reset << value << " not in range [" << min_val << ", " << max_val << "]" << std::endl;
-                    LOG( WARNING ) << value << " not in range [" << min_val << ", " << max_val << "]" << std::endl;
+                    LOG( WARNING ) << value << " not in range [" << min_val << ", " << max_val << "] for parameter named " << name << std::endl;
```
please use fmt::fornat

prudhomm commented on mor/mor/feel/feelmor/parameterspace.hpp:
```diff
@@ -220,8 +220,8 @@ class ParameterSpace: public CommObject, public std::enable_shared_from_this<Par
-                    Feel::cout << tc::red << "Warning: " << tc::reset << value << " not in range [" << min_val << ", " << max_val << "]" << std::endl;
-                    LOG( WARNING ) << value << " not in range [" << min_val << ", " << max_val << "]" << std::endl;
+                    LOG( WARNING ) << value << " not in range [" << min_val << ", " << max_val << "] for parameter named " << name << std::endl;
+                    throw std::invalid_argument("Parameter out of range");
```
please use a similar message as the line above using fmt::format to produce the string

prudhomm commented on mor/mor/feel/feelmor/parameterspace.hpp:
```diff
@@ -235,8 +235,8 @@ class ParameterSpace: public CommObject, public std::enable_shared_from_this<Par
             {
-                Feel::cout << tc::red << "Warning: " << tc::reset << value << " not in range [" << min(i) << ", " << max(i) << "]" << std::endl;
-                LOG( WARNING ) << value << " not in range [" << min(i) << ", " << max(i) << "]" << std::endl;
+                LOG( WARNING ) << value << " not in range [" << min(i) << ", " << max(i) << "] for parameter number " << i << std::endl;
```
@thomas-saigre see above

prudhomm commented on mor/mor/feel/feelmor/parameterspace.hpp:
```diff
@@ -235,8 +235,8 @@ class ParameterSpace: public CommObject, public std::enable_shared_from_this<Par
-                Feel::cout << tc::red << "Warning: " << tc::reset << value << " not in range [" << min(i) << ", " << max(i) << "]" << std::endl;
-                LOG( WARNING ) << value << " not in range [" << min(i) << ", " << max(i) << "]" << std::endl;
+                LOG( WARNING ) << value << " not in range [" << min(i) << ", " << max(i) << "] for parameter number " << i << std::endl;
+                throw std::invalid_argument("Parameter out of range");
```
@thomas-saigre see above

thomas-saigre committed:
use fmt::format instead of <<
/cc @prudhomm
[skip feelpp, skip toolboxes, skip tests, skip python]

thomas-saigre committed:
up doc (and relaunch pipeline) #2074

thomas-saigre committed:
handle exceptions in tests
/cc @prudhomm
[skip feelpp, skip tests, skip mor, skip toolboxes]

thomas-saigre committed:
fix test execution
[skip feelpp, skip toolboxes, skip tests, skip mor]
